src/ccl_power.c

```diff
@@ -489,71 +488,80 @@ static void ccl_cosmology_compute_power_class(ccl_cosmology * cosmo, int * statu
-    }
-
-    //These are the limits of the splining range
-    cosmo->data.k_min_nl=2*exp(sp.ln_k[0]);
-    cosmo->data.k_max_nl=ccl_splines->K_MAX_SPLINE;
-
-    if(cosmo->config.matter_power_spectrum_method==ccl_halofit) {
-
-      double psout_nl;
+      return;
+    } else {
 
-      for (int i=0; i<nk; i++) {
-	for (int j = 0; j < na; j++) {
-	  s |= spectra_pk_nl_at_k_and_z(&ba, &pm, &sp,exp(x[i]),1./z[j]-1.,&psout_nl);
-	  y2d_nl[j*nk+i] = log(psout_nl);
+      //These are the limits of the splining range
+      cosmo->data.k_min_nl=2*exp(sp.ln_k[0]);
+      cosmo->data.k_max_nl=ccl_splines->K_MAX_SPLINE;
+      
+      if(cosmo->config.matter_power_spectrum_method==ccl_halofit) {
+	
+	double psout_nl;
+	
+	for (int i=0; i<nk; i++) {
+	  for (int j = 0; j < na; j++) {
+	    s |= spectra_pk_nl_at_k_and_z(&ba, &pm, &sp,exp(x[i]),1./a[j]-1.,&psout_nl);
+	    y2d_nl[j*nk+i] = log(psout_nl);
+	  }
+	}
+	
+	if(s) {
+	  free(x);
+	  free(a);
+	  free(y2d_nl);
+	  free(y2d_lin);
+	  *status = CCL_ERROR_CLASS;
+	  ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error computing CLASS power spectrum\n");
+	  ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
+	  return;
 	}
-      }
-
-      if(s) {
-	free(x);
-	free(z);
-	free(y2d_nl);
-	free(y2d_lin);
-	*status = CCL_ERROR_CLASS;
-	ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error computing CLASS power spectrum\n");
-	ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
-	return;
-
-      }
 
-      gsl_spline2d * log_power_nl = gsl_spline2d_alloc(PNL_SPLINE_TYPE, nk,na);
-      pwstatus = gsl_spline2d_init(log_power_nl, x, z, y2d_nl,nk,na);
-
-      if (pwstatus) {
-	free(x);
-	free(z);
-	free(y2d_nl);
-	free(y2d_lin);
-	gsl_spline2d_free(log_power_nl);
-	ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
-	*status = CCL_ERROR_SPLINE;
-	ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error creating log_power_nl spline\n");
-	return;
-      }
-      else {
-	cosmo->data.p_nl = log_power_nl;
+	gsl_spline2d * log_power_nl = gsl_spline2d_alloc(PNL_SPLINE_TYPE, nk,na);
+	pwstatus = gsl_spline2d_init(log_power_nl, x, a, y2d_nl,nk,na);
+	
+	if (pwstatus) {
+	  free(x);
+	  free(a);
+	  free(y2d_nl);
+	  free(y2d_lin);
+	  gsl_spline2d_free(log_power_nl);
+	  ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
+	  *status = CCL_ERROR_SPLINE;
+	  ccl_cosmology_set_status_message(cosmo, "ccl_power.c: ccl_cosmology_compute_power_class(): Error creating log_power_nl spline\n");
+	  return;
+	} else {
+	  cosmo->data.p_nl = log_power_nl;
+	}
       }
-
+      
+      ccl_free_class_structs(cosmo, &ba,&th,&pt,&tr,&pm,&sp,&nl,&le,init_arr,status);
+      free(x);
+      free(y2d_lin);
+      free(a);
       free(y2d_nl);
```

Comment:
I find all these frees all over the place hard to follow.  I still recommend switching over to the pattern that Tom has implemented in #507.  It makes it much clearer that everything is properly freed, since you do them all only once just before returning (also only once).

(Fine though if the plan is to do more refactorings like that in other PRs.)

Reply:
I changed `ccl_cosmology_compute_power_class` following the example of #507. Let me know if this is what you meant, then I can refactor the rest.